pkg/controller/agent/controller.go

```diff
@@ -144,22 +147,35 @@ func (r *ReconcileAgent) Reconcile(ctx context.Context, request reconcile.Reques
 		return reconcile.Result{}, nil
 	}
 
-	res, err := r.doReconcile(ctx, agent).Aggregate()
-	k8s.EmitErrorEvent(r.recorder, err, &agent, events.EventReconciliationError, "Reconciliation error: %v", err)
+	result, err = r.doReconcile(ctx, *agent).Aggregate()
+	k8s.EmitErrorEvent(r.recorder, err, agent, events.EventReconciliationError, "Reconciliation error: %v", err)
 
-	return res, err
+	return result, err
 }
 
-func (r *ReconcileAgent) doReconcile(ctx context.Context, agent agentv1alpha1.Agent) *reconciler.Results {
+func (r *ReconcileAgent) doReconcile(ctx context.Context, agent agentv1alpha1.Agent) (results *reconciler.Results) {
 	defer tracing.Span(&ctx)()
-	results := reconciler.NewResult(ctx)
+	var err error
+	results = reconciler.NewResult(ctx)
+	status := newStatus(agent)
+
+	// defer the updating of status to ensure that the status is updated regardless of the outcome of the reconciliation.
+	// note that this deferred function is modifying the return values, which are named return values, which allows this
+	// to function properly.
+	defer func() {
+		if updateStatusresults := updateStatus(ctx, agent, r.Client, status).WithError(err); updateStatusresults != nil {
```

Comment:
It is not clear to me why we want to call `WithError` here with variable `err` we closed over in the defer. Either that variable will be `nil` in which case the `WithError` call is redundant, or the variable points to an error in which case we  adding the same error twice as we already called `WithError` for example on L177 (which is the right thing to do). 

Instead I would close over the `results` variable and pass it in to `updateStatus` so you don't have to create. new `result` and you preserve all the results and errors that have potentially been produced further down the call stack. In your current implementation you will lose any error or result returned from functions from `doReconcile` and you only capture the `err` variable in this function. If you don't want to pass in the results into `updateStatus` you need to at least merge them with the results produced by the main function body.